### PR TITLE
Introduce smoke test

### PR DESCRIPTION
**Pull request by steveklabnik**
This checks that we render the example book correctly. In order to do
that, we put a known-good copy of the book under tests/book, and then in
tests/smoke.rs, we generate a copy of it, and then diff the two
directories.

This means that PRs that change the generated output will need to update
this fixture, but it also means we get to see an easy way of what they
actually update.

Part of #11

### CODE REVIEW

**Comment by azerupi** [comment]
Oh yes yes yes! I am all in for better tests :)
I have accepted a couple of PRs so make sure that the test passed before merging.

**Comment by steveklabnik** [comment]
>  I have accepted a couple of PRs so make sure that the test passed before merging.

👍 ❤️

**Comment by steveklabnik** [comment]
only a small amount of change, as you can see in the second commit 😄

**Comment by steveklabnik** [comment]
Ahh, line endings are the issue here.... I made this PR on Windows, so windows stuff succeeds, but linux stuff fails. Hrm.

**Comment by steveklabnik** [comment]
..... I am not sure how to fix this, given that if I change it the other way, it will fail on Windows. Hrm.

**Comment by carols10cents** [comment]
Sounds like something that should be fixed in dir-diff, I [filed an issue for you](https://github.com/steveklabnik/dir-diff/issues/9) ❤️

**Comment by steveklabnik** [comment]
:heart:

File: tests/book/index.html [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+                    <h1 class="menu-title">mdBook Documentation</h1>
+
+                    <div class="right-buttons">
+                        <i id="print-button" class="fa fa-print" title="Print this book"></i>
+                    </div>
+                </div>
+
+                <div id="content" class="content">
+                    <a class="header" href="misc\introduction.html#introduction" id="introduction"><h1>Introduction</h1></a>
```
Comment (ollie27) [inline tests/book/index.html]: This is the actual problem. Backslashes shouldn't be used in URLs. I've opened a more relevant issue: #263

**Comment by azerupi** [comment]
Thinking about this a bit, this kind of test might be a bit too restrictive. Since the output often changes with small PRs I think it would be better to drop this idea and instead implement a lot of small tests checking for one particular property in the generated output. It would probably not catch everything, but it would be more granular than this solution. 

Now the major drawback would be to actually write all of the small tests. But we have had a lot of contributions recently and I think that if we make easy issues, we can get most of them contributed as mentored PRs.

What do you think?

**Comment by Michael-F-Bryan** [comment]
Hey @azerupi and @steveklabnik I just noticed this PR.

> I think it would be better to drop this idea and instead implement a lot of small tests checking for one particular property in the generated output

I've started something similar in #374. In particular the `/tests/rendered_output.rs` test file seems to do exactly what you are talking about. If I can abstract the checking out into a nice macro, would it be worth merging that and then supplementing it with a "fuzzier" version of this smoke test (perhaps something which compares simplified HTML ASTs)?

I'd also be happy to mentor a PR for asserting particular patterns appear in the output as expected.

**Comment by steveklabnik** [comment]
Let's do it! 😄

**Pull request closed without merging**